server/src/test/java/org/elasticsearch/action/search/SearchResponseMergerTests.java

```diff
@@ -0,0 +1,435 @@
+        final long currentRelativeTime = randomLong();
+        final SearchTimeProvider timeProvider = new SearchTimeProvider(randomLong(), 0, () -> currentRelativeTime);
+        final SearchResponse.Clusters clusters = SearchResponseTests.randomClusters();
+        final int size = randomIntBetween(0, 100);
+        final int from = size > 0 ? randomIntBetween(0, 100) : 0;
+        final int requestedSize = from + size;
+        final SortField[] sortFields;
+        final String collapseField;
+        boolean scoreSort = false;
+        if (randomBoolean()) {
+            int numFields = randomIntBetween(1, 3);
+            sortFields = new SortField[numFields];
+            for (int i = 0; i < numFields; i++) {
+                final SortField sortField;
+                if (randomBoolean()) {
+                    sortField = new SortField("field-" + i, SortField.Type.INT, randomBoolean());
+                } else {
+                    scoreSort = true;
+                    sortField = SortField.FIELD_SCORE;
+                }
+                sortFields[i] = sortField;
+            }
+            collapseField = randomBoolean() ? "collapse" : null;
+        } else {
+            collapseField = null;
+            sortFields = null;
+            scoreSort = true;
+        }
+        TotalHits.Relation totalHitsRelation = frequently() ? randomFrom(TotalHits.Relation.values()) : null;
+
+        PriorityQueue<SearchHit> priorityQueue = new PriorityQueue<>(new SearchHitComparator(sortFields));
+        SearchResponseMerger searchResponseMerger = new SearchResponseMerger(from, size, timeProvider, clusters, flag -> null);
+        TotalHits expectedTotalHits = null;
+        int expectedTotal = 0;
+        int expectedSuccessful = 0;
+        int expectedSkipped = 0;
+        int expectedReducePhases = 1;
+        boolean expectedTimedOut = false;
+        Boolean expectedTerminatedEarly = null;
+        float expectedMaxScore = Float.NEGATIVE_INFINITY;
+        int numIndices = requestedSize == 0 ? 0 : randomIntBetween(1, requestedSize);
+        Iterator<Map.Entry<String, Index[]>> indicesIterator = randomRealisticIndices(numIndices, numResponses).entrySet().iterator();
+        for (int i = 0; i < numResponses; i++) {
+            Map.Entry<String, Index[]> entry = indicesIterator.next();
+            String clusterAlias = entry.getKey().equals(RemoteClusterAware.LOCAL_CLUSTER_GROUP_KEY) ? null : entry.getKey();
+            Index[] indices = entry.getValue();
+            int total = randomIntBetween(1, 1000);
+            expectedTotal += total;
+            int successful = randomIntBetween(1, total);
+            expectedSuccessful += successful;
+            int skipped = randomIntBetween(1, total);
+            expectedSkipped += skipped;
+
+            TotalHits totalHits = null;
+            if (totalHitsRelation != null) {
+                //TODO totalHits may overflow if each cluster reports a very high number?
```

Comment:
@jimczi do you have thoughts on this one? is it paranoia on my end?

Reply:
totalHits.value is a long so I doubt that it will overflow unless you have millions of remote cluster ;). However this made me think that this pr doesn't handle `track_total_hits` when it is set as a number. In the normal execution we'll merge all the topdocs and if the resulting total hits is greater than `track_total_hits` we set the final value as the value in `track_total_hits` and the final relation to `gte`. You can check the logic here https://github.com/elastic/elasticsearch/blob/95479f17667c7d27fc480c260a1c380d681d8b93/server/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java#L759.
We'll also need to change the code when https://github.com/elastic/elasticsearch/pull/37466 is merged since the default for `track_total_hits` is going to change.

Reply:
Good point, I addressed that in a3251291806a1d46a3c9e4f7d95779b5460b80c6 . I think that since I reuse `TopDocsStats`, this will not need chaning once the default changes.

I also went a step further in 8f1b063e0a3516e358caf377c19371ac3e8cd8ff and expanded TopDocsStats. Just an experiment but it saves some duplicated logic, let me know what you think.

Reply:
thanks, the change looks good. Feel free to push when CI is green